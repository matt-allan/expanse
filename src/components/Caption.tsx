import React, { useEffect, useState } from "react";
import { Box, Text } from "grommet";

type Frame = {
  start: number;
  end: number;
  caption: string;
};

type CaptionProps = {
  frames: Frame[];
};

export const Caption = ({ frames }: CaptionProps): JSX.Element | null => {
  const [seconds, setSeconds] = useState(0);

  useEffect(() => {
    const interval = setInterval(() => {
      setSeconds((prev) => prev + 1);
    }, 1000);

    return () => clearInterval(interval);
  }, []);

  for (const frame of frames) {
    if (frame.start <= seconds && frame.end >= seconds) {
      return (
        <Box
          animation={[
            { type: "fadeIn", delay: frame.start * 1000, duration: 1000 },
            { type: "fadeOut", delay: frame.end * 1000 - 500, duration: 500 },
          ]}
        >
          <Text>{frame.caption}</Text>
        </Box>
      );
    }
  }

  return null;
};
