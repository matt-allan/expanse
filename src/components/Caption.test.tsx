import React from "react";
import { Caption } from "./Caption";
import { create } from "react-test-renderer";

jest.useFakeTimers();

it("renders", () => {
  const component = create(
    <Caption
      frames={[
        { start: 0, end: 3, caption: "Hello..." },
        { start: 4, end: 7, caption: "World" },
      ]}
    />
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
