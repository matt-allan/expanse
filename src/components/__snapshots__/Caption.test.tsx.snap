// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 dMKPgk"
>
  <span
    className="StyledText-sc-1sadyjn-0 hUokoe"
  >
    Hello...
  </span>
</div>
`;
